grants_tagger/models/mesh_xlinear.py

```diff
@@ -25,70 +27,120 @@ def __init__(
+
         logger.info("Fitting vectorizer")
-        if not hasattr(self, "vectorizer"):
-            self._init_vectorizer()
-        self.vectorizer.fit(X)
 
-        X_vec = self.vectorizer.transform(X).astype("float32")
+        self._init_vectorizer()
+
+        if self.vectorizer_library == "sklearn":
+            X_vec = self.vectorizer_.fit_transform(X).astype("float32")
+        else:
+            self.vectorizer_ = self.vectorizer_.train(
+                X,
+                config={
+                    "ngram_range": self.ngram_range,
+                    "max_feature": self.max_features,
+                    "min_df_cnt": self.min_df,
+                },
+            )
+            X_vec = self.vectorizer_.predict(X).astype("float32")
+
         Y = Y.astype("float32")
 
         logger.info("Creating cluster chain")
+
         label_feat = LabelEmbeddingFactory.create(Y, X_vec, method="pifa")
         cluster_chain = Indexer.gen(label_feat, indexer_type="hierarchicalkmeans")
+        xlinear_model = XLinearModel()
 
         logger.info("Training model")
-        model = XLinearModel()
-        self.model = model.train(X_vec, Y, C=cluster_chain, **self.model_params)
+
+        # Sklearn estimators need variables introduced during training to have a trailing underscore
+        self.xlinear_model_ = xlinear_model.train(
+            X_vec,
+            Y,
+            C=cluster_chain,
+            cluster_chain=self.cluster_chain,
+            negative_sampling_scheme=self.negative_sampling_scheme,
+            only_topk=self.only_topk,
+            threshold=self.min_weight_value,
+        )
         return self
 
     def predict(self, X):
         return self.predict_proba(X) > 0.5
 
     def predict_proba(self, X):
-        return self.model.predict(self.vectorizer.transform(X).astype("float32"))
+        if self.vectorizer_library == "sklearn":
+            return self.xlinear_model_.predict(
+                self.vectorizer_.transform(X).astype("float32"),
+                beam_size=self.beam_size,
+            )
+        else:
+            return self.xlinear_model_.predict(
+                self.vectorizer_.predict(X).astype("float32"), beam_size=self.beam_size
+            )
 
     def save(self, model_path):
         vectorizer_path = os.path.join(model_path, "vectorizer.pkl")
-        save_pickle(vectorizer_path, self.vectorizer)
-        self.model.save(model_path)
+        if self.vectorizer_library == "sklearn":
+            save_pickle(vectorizer_path, self.vectorizer_)
+        else:
+            self.vectorizer_.save(model_path)
+
+        self.xlinear_model_.save(model_path)
 
     def load(self, model_path, is_predict_only=True):
         vectorizer_path = os.path.join(model_path, "vectorizer.pkl")
-        self.vectorizer = load_pickle(vectorizer_path)
-        self.model = XLinearModel.load(model_path, is_predict_only=is_predict_only)
+        if self.vectorizer_library == "sklearn":
```

Comment:
not that this is a bad strategy but what would be even better is to save the params in a `meta.json` and initialize from there so that you don't need to pass params when loading the model. otherwise a user might save a vectorizer and then try to load without knowing which vectorizer was used. the information should ideally be stored somewhere. `spacy` uses a `meta.json` as well. I have short of tried following that route but it is not complete yet in other models so its optional

Reply:
Yep, meta route is definitely better. Let's add this to improvements!